### DISCUSSION TITLE
Moving from flux v1 to flux v2 with bootstrap possibility

### DISCUSSION THREAD

**Discussion by bhavyapandey** [fluxcd/flux2#4417] [General] [answered]
I'm trying to move from flux v1 to flux2 on a lot of our prod and non prod cluster. 
I'm new to this setup and tryig to understand better.
With that said, flux v1 was installed using the 'kubectl apply' way.. 
Though updating it in a similar fashion is straight-forward and I'm able to achieve it. 
I am looking for ways to achieve the gitops way, like flux bootstrap.. 
I don't want to delete my old installation, just upgrade and configure bitbucket to have the gotk-components.yaml and then flux is able to maintain itself. 
I've looked in the documentation but found no way. 
Is this possible?

**Comment by kingdonb** [comment, marked as answer]
If your Flux v1 was installed via "kubectl apply -f" then presumably you still have the definitions from which it was applied, hopefully they are in git? The same method is still available in Flux v2, `flux install` which is effectively the "kubectl apply" way, you are right to want to do it better and do the GitOps way instead!

The Flux v2 is a total rewrite from Flux v1, they share little code in common if any, the configuration is totally different. We have a migration guide, there are a list of guides on this page: https://fluxcd.io/flux/migration/

Flux v2 is installed by default in a different namespace than Flux v1, you will end up deleting your old installation, there is no direct upgrade because they are different products in effect. You can run both Flux v1 and Flux v2 concurrently, but there is no really good reason to do this for very long. And you will hit problems if you have Flux v1 managing the same resources/files as Flux v2, so try to avoid doing that at all costs. It should be easy to disable Flux v1 and bootstrap Flux v2 at the same time, so you can delete the old manifests. (Is there a reason you want to keep them around?)

There is no "just upgrade and everything is the same" unfortunately, we have seen a lot of users find some struggle because of the different way that Flux v2 applies resources. We think this way is better, but there are some new concepts applied like server-side reconciling and atomic reconciliation, Custom Resource Definitions, error handling and health checking; there are some overviews that vary in complexity to help you understand. I would refer to the Flux E2E guide and to the Core Concepts, probably not in that order. Flux E2E is the most complex doc, the Migration guide is somewhat complex, I would start with the simplest docs first. (From the top!)

The migration guide is probably what you are looking for, but if you have some issue, you are welcome to post questions here or in the Flux channel on the CNCF slack, where people are also quite responsive and helpful, and everyone who supports Flux is doing so as volunteers. I would suggest reading these docs in this order first:

* https://fluxcd.io/flux/concepts/
* https://fluxcd.io/flux/get-started/
* https://fluxcd.io/flux/installation/ -- has different sections depending on your specific environment
* https://fluxcd.io/flux/migration/flux-v1-migration/
* https://fluxcd.io/flux/flux-e2e/
* Several docs related to Flux and Helm, depending on your use case

There is a sub-section of the installation guide with specifics for Bitbucket Cloud, so check it out for more details, and hope this helps!
